### PR TITLE
Make mount type optional for bind mounts

### PR DESCRIPTION
**Pull request by moricho**
## Description
This fixes behavior of `getMountNameAndOptions` func based on OCI runtime spec(https://github.com/opencontainers/runtime-spec/blob/master/config.md#posix-platform-mounts).
In detail, if the mount type is equal to "none" or other dummy string and the mount option includes either "bind" or  "rbind",  `getMountNameAndOptions` behave as the mount type is equal to "bind".

## Relevant issue
Fix: https://github.com/google/gvisor/issues/2330

Signed-off-by: moricho <ikeda.morito@gmail.com>

### CODE REVIEW

File: runsc/boot/fs.go [hunk 1]
```diff
@@ -756,6 +767,17 @@ func (c *containerMounter) createRootMount(ctx context.Context, conf *Config) (*
...
+	fsName := "9p"
+	opts := p9MountOptions(fd, c.getMountAccessType(m))
+	// If configured, add overlay to all writable mounts.
+	useOverlay := conf.Overlay && !mountFlags(m.Options).ReadOnly
+	return fsName, opts, useOverlay
+}
+
 // getMountNameAndOptions retrieves the fsName, opts, and useOverlay values
 // used for mounts.
 func (c *containerMounter) getMountNameAndOptions(conf *Config, m specs.Mount) (string, []string, bool, error) {
```
Comment (fvoznika) [inline runsc/boot/fs.go:783]: I think you can contain the implementation entirely inside this function. First, check that options `bind` or `mbind` are present. If yes, you overwrite m.Type to `bind` and run the switch below as is. The rest of the code should just work.

Please add a a simple test covering this case in container_test.go.

File: runsc/boot/fs.go [hunk 2]
```diff
@@ -219,6 +224,8 @@ func mountFlags(opts []string) fs.MountSourceFlags {
 			mf.NoAtime = true
 		case "noexec":
 			mf.NoExec = true
+		case "bind", "rbind":
+			mf.Bind = true
```
Comment (fvoznika) [inline runsc/boot/fs.go:228]: In this case you don't need to set anything to `mf`. Just add a comment that "bind" in options is converted to a 9P mount.

File: runsc/boot/fs.go [hunk 3]
```diff
@@ -765,6 +776,15 @@ func (c *containerMounter) getMountNameAndOptions(conf *Config, m specs.Mount) (
 		useOverlay bool
 	)
 
+	for _, opt := range m.Options {
+		// When options include either "bind" or "rbind", this behaves as bind mount.
+		// However, if mount type is equal to a filesystem supported on runsc, this prioritizes it over bind mount.
```
Comment (fvoznika) [inline runsc/boot/fs.go:781]: Please keep comment lines within 80 columns.

File: runsc/boot/fs.go [hunk 4]
```diff
@@ -765,6 +776,15 @@ func (c *containerMounter) getMountNameAndOptions(conf *Config, m specs.Mount) (
...
+		if (opt == "bind" || opt == "rbind") && !specutils.ContainsStr(supportedFilesystems, m.Type) {
```
Comment (fvoznika) [inline runsc/boot/fs.go:782]: According to the spec, type can be ignored for bind mounts. So you don't need to check for `supportedFilesystems`.

> For bind mounts (when options include either bind or rbind), the type is a dummy, often "none"

File: runsc/boot/fs.go [hunk 5]
```diff
@@ -778,14 +798,12 @@ func (c *containerMounter) getMountNameAndOptions(conf *Config, m specs.Mount) (
 		if err != nil {
 			return "", nil, false, err
 		}
-
```
Comment (fvoznika) [inline runsc/boot/fs.go:781]: Please keep the extra line.

File: runsc/boot/fs.go [hunk 6]
```diff
@@ -778,14 +798,12 @@ func (c *containerMounter) getMountNameAndOptions(conf *Config, m specs.Mount) (
...
 	case bind:
 		fd := c.fds.remove()
 		fsName = "9p"
 		opts = p9MountOptions(fd, c.getMountAccessType(m))
 		// If configured, add overlay to all writable mounts.
 		useOverlay = conf.Overlay && !mountFlags(m.Options).ReadOnly
-
```
Comment (fvoznika) [inline runsc/boot/fs.go:788]: Please keep the extra line.

File: runsc/boot/fs.go [hunk 7]
```diff
@@ -230,6 +237,10 @@ func isSupportedMountFlag(fstype, opt string) bool {
 	switch opt {
 	case "rw", "ro", "noatime", "noexec":
 		return true
+	case "bind", "rbind":
+		if fstype == nonefs || !specutils.ContainsStr(supportedFilesystems, fstype) {
+			return true
+		}
```
Comment (fvoznika) [inline runsc/boot/fs.go:243]: This check is not needed, because these options do not change mount behavior. Please remove.

**Comment by moricho** [comment]
@fvoznika Thanks for your review! I fixed.

**Pull request merged**